tsuru/client/plugin.go

```diff
@@ -57,37 +62,208 @@ func (c *PluginInstall) Run(context *cmd.Context, client *cmd.Client) error {
 	pluginName := context.Args[0]
 	pluginURL := context.Args[1]
 	if err := installPlugin(pluginName, pluginURL); err != nil {
-		return err
+		return fmt.Errorf("Error installing plugin %q: %w", pluginName, err)
 	}
 
 	fmt.Fprintf(context.Stdout, `Plugin "%s" successfully installed!`+"\n", pluginName)
 	return nil
 }
 
 func installPlugin(pluginName, pluginURL string) error {
-	pluginPath := cmd.JoinWithUserDir(".tsuru", "plugins", pluginName)
-	file, err := filesystem().OpenFile(pluginPath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0755)
+	tmpDir, err := filesystem().MkdirTemp("", "")
 	if err != nil {
-		return err
+		return fmt.Errorf("Could not create a tmpdir: %w", err)
 	}
+	defer filesystem().RemoveAll(tmpDir)
+
 	resp, err := http.Get(pluginURL)
 	if err != nil {
-		return err
+		return fmt.Errorf("Could not GET %q: %w", pluginURL, err)
 	}
+
 	defer resp.Body.Close()
-	data, err := ioutil.ReadAll(resp.Body)
+
+	data, err := io.ReadAll(resp.Body)
 	if err != nil {
 		return err
 	}
 	if resp.StatusCode < 200 || resp.StatusCode >= 400 {
 		return fmt.Errorf("Invalid status code reading plugin: %d - %q", resp.StatusCode, string(data))
 	}
-	n, err := file.Write(data)
+
+	// Try to extract .tar.gz first, then .zip. Fallbacks to copy the content
+	extractErr := extractTarGz(tmpDir, bytes.NewReader(data))
+	if extractErr != nil {
+		extractErr = extractZip(tmpDir, bytes.NewReader(data))
+	}
+	if extractErr != nil {
+		file, err := filesystem().OpenFile(filepath.Join(tmpDir, pluginName), os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0755)
+		if err != nil {
+			return fmt.Errorf("Failed to open file: %w", err)
+		}
+		defer file.Close()
+		n, err := file.Write(data)
+		if err != nil {
+			return fmt.Errorf("Failed to write file content: %w", err)
+		}
+		if n != len(data) {
+			return fmt.Errorf("Incomplete write")
+		}
+	}
+
+	executablePath := findExecutablePlugin(tmpDir, pluginName)
+	if executablePath == "" {
+		return fmt.Errorf("The downloaded plugin content is invalid.")
+	}
+
+	pluginPath := cmd.JoinWithUserDir(".tsuru", "plugins", pluginName)
+	if extractErr == nil {
+		os.Chmod(tmpDir, 0755)
+		if _, err := filesystem().Stat(pluginPath); err == nil {
+			filesystem().RemoveAll(pluginPath)
+		}
+		if err := filesystem().Rename(tmpDir, pluginPath); err != nil {
+			return fmt.Errorf("Could not move tmpDir: %w", err)
+		}
+	} else {
+		if err := copyFile(executablePath, pluginPath); err != nil {
+			return fmt.Errorf("Could not write plugin file: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func findExecutablePlugin(basePath, pluginName string) (execPath string) {
+	testPathGlobs := []string{
+		filepath.Join(basePath, pluginName),
+		filepath.Join(basePath, pluginName, pluginName),
+		filepath.Join(basePath, pluginName, pluginName+".*"),
+		filepath.Join(basePath, pluginName+".*"),
```

Comment:
I though about the following scenario:
- user has the `myplugin.py` 
- user updates the plugin (which is using the newer method)
- the update (plugin install) which would create `myplugin/myplugin.py`

Following the order as you suggested, you would need to manually remove the old plugin (and it would not be obvious to the user).

On the other hand, if you currently have both `myscript/myscript.fake` and `myscript.py`, this update would break your plugin. 

I believe the first scenario to be far likely than the 2nd, so I preferred my approach 
(despite both being very unlikely in my opinion)

Do you agree?

Reply:
I think I got the problem. This function is used in both cases: 1) to select the executable from a zipped file; 2) to select the executable from `~/.tsuru/plugins` directory. I thought only in the former case when I suggested it. So, if there a plugin as below we should match `my-plugin.py` first.

```
- my-plugin.tar.gz
  +  dir1
    - my-plugin.py
  +  my-plugin.py
```

For the later I agree w/ you. Despite shouldn't exist a plugin w/ extension within `~/.tsuru/plugins` dir, at least while installing the plugin via `tsuru plugin install`.

Reply:
Yup. I'm only matching for the first sub-dir.
I'm only creating a dir for a compacted resource (zip | tar.gz)

---

## Some examples:

executes: `~/.tsuru/plugins/myplugin/myplugin.sh` (instead of the old `~/.tsuru/plugins/myplugin.py`)
```
~/.tsuru
└── plugins
    ├── myplugin
    │   ├── myplugin
    │   │   ├── myplugin
    │   │   └── myplugin.sh
    │   └── myplugin.sh
    └── myplugin.py
```

executes: `~/.tsuru/plugins/myplugin/myplugin.sh`
```
~/.tsuru
└── plugins
    └── myplugin
        ├── myplugin
        │   ├── myplugin
        │   └── myplugin.sh
        └── myplugin.sh
```

This won't work, the cli will fail to install or execute
```
~/.tsuru
└── plugins
    └── myplugin
        └── myplugin
            ├── myplugin
            └── myplugin.sh
```

Reply:
So much clear now. Thanks @tcarreira! :heart: